### PR TITLE
BUG: Inconsistent conversion of missing column names

### LINKED ISSUE #44818: BUG: Inconsistent conversion of missing column names
### 

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the master branch of pandas.

### Reproducible Example

```python
import pandas as pd
f = pd.DataFrame([1], index=pd.MultiIndex.from_arrays([[2],[3],[4]], names=[None, "a", None]))

>>> f.reset_index()
   level_0  a  level_2  0
0        2  3        4  1

>>> pd.DataFrame(f.to_records())
   level_0  a  level_1  0
0        2  3        4  1

>>> f.index.to_frame(index=False)
   0  a  2
0  2  3  4
```

### Issue Description

`DataFrame.to_records()` is dealing with blank column labels by counting occurences and prefixing with "level_"
`Index `and `MultiIndex.to_frame()` is using the un-prefixed column count.
The rest of Pandas uses the column count prefixed with "level_"

### Expected Behavior

```
>>> pd.DataFrame(f.to_records())
   level_0  a  level_2  0
0        2  3        4  1

>>> f.index.to_frame(index=False)
   level_0  a  level_2
0        2  3        4

```

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 04b538a55365b5d1ae930687a91bb3ac4ea53040
python           : 3.8.12.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.19043
machine          : AMD64
processor        : Intel64 Family 6 Model 158 Stepping 9, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : English_United Kingdom.1252

pandas           : 1.4.0.dev0+1332.g04b538a553
numpy            : 1.21.4
pytz             : 2021.3
dateutil         : 2.8.2
pip              : 21.3.1
setuptools       : 59.2.0
Cython           : 0.29.24
pytest           : 6.2.5
hypothesis       : 6.25.0
sphinx           : 4.3.0
blosc            : None
feather          : None
xlsxwriter       : 3.0.2
lxml.etree       : 4.6.4
html5lib         : 1.1
pymysql          : None
psycopg2         : None
jinja2           : 3.0.3
IPython          : 7.29.0
pandas_datareader: None
bs4              : 4.10.0
bottleneck       : 1.3.2
fsspec           : 2021.11.0
fastparquet      : 0.7.1
gcsfs            : 2021.11.0
matplotlib       : 3.5.0
numexpr          : 2.7.3
odfpy            : None
openpyxl         : 3.0.9
pandas_gbq       : None
pyarrow          : 6.0.1
pyxlsb           : None
s3fs             : 2021.11.0
scipy            : 1.7.2
sqlalchemy       : 1.4.27
tables           : 3.6.1
tabulate         : 0.8.9
xarray           : 0.18.2
xlrd             : 2.0.1
xlwt             : 1.3.0
numba            : 0.53.1

</details>

### PR DESCRIPTION
**Pull request by johnzangwill**
- [x] closes #44818
- [x] tests changed / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

Factored out multiple occurances of missing name logic into common method.
DataFrame.to_records() changed to use common method rather than missing value count when filling missing names.
Index.to_frame() left alone for the time being.
```
import pandas as pd
f = pd.DataFrame([1], index=pd.MultiIndex.from_arrays([[2],[3],[4]], names=[None, "a", None]))
```
Old behavior: 
```
>>> pd.DataFrame(f.to_records())
   level_0  a  level_1  0
0        2  3        4  1
```
New  behavior:
```
>>> pd.DataFrame(f.to_records())
   level_0  a  level_2  0
0        2  3        4  1
```

### CODE REVIEW

File: pandas/tests/frame/methods/test_to_records.py [hunk 1]
```diff
@@ -91,7 +91,7 @@ def test_to_records_index_name(self):
         df.index = MultiIndex.from_tuples([("a", "x"), ("a", "y"), ("b", "z")])
         df.index.names = ["A", None]
         rs = df.to_records()
-        assert "level_0" in rs.dtype.fields
+        assert "level_1" in rs.dtype.fields
```
Comment (phofl) [inline pandas/tests/frame/methods/test_to_records.py:94]: Could you maybe add a test checking the whole result?

File: doc/source/whatsnew/v1.4.0.rst [hunk 2]
```diff
@@ -854,6 +854,7 @@ Other
 - Bug in :meth:`DataFrame.shift` with ``axis=1`` and ``ExtensionDtype`` columns incorrectly raising when an incompatible ``fill_value`` is passed (:issue:`44564`)
 - Bug in :meth:`DataFrame.diff` when passing a NumPy integer object instead of an ``int`` object (:issue:`44572`)
 - Bug in :meth:`Series.replace` raising ``ValueError`` when using ``regex=True`` with a :class:`Series` containing ``np.nan`` values (:issue:`43344`)
+- Bug in :meth:`DataFrame.to_records` missing names filled incorrectly (:issue:`44818`)
```
Comment (phofl) [inline doc/source/whatsnew/v1.4.0.rst:857]: Could you specify a bit more?

**Review by phofl — COMMENTED** [review]
small comments, otherwise lgtm

File: doc/source/whatsnew/v1.4.0.rst (hunk 2 above)
Comment (johnzangwill) [reply doc/source/whatsnew/v1.4.0.rst]: Done

**Review by phofl — APPROVED** [review]

File: doc/source/whatsnew/v1.4.0.rst [hunk 3]
```diff
@@ -856,6 +856,7 @@ Other
 - Bug in :meth:`DataFrame.shift` with ``axis=1`` and ``ExtensionDtype`` columns incorrectly raising when an incompatible ``fill_value`` is passed (:issue:`44564`)
 - Bug in :meth:`DataFrame.diff` when passing a NumPy integer object instead of an ``int`` object (:issue:`44572`)
 - Bug in :meth:`Series.replace` raising ``ValueError`` when using ``regex=True`` with a :class:`Series` containing ``np.nan`` values (:issue:`43344`)
+- Bug in :meth:`DataFrame.to_records` where an incorrect n was used when missing names were replaced by level_n (:issue:`44818`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.4.0.rst:859]: can you add backticks around the ``n`` and ``level_n``

File: pandas/core/common.py [hunk 4]
```diff
@@ -604,3 +604,10 @@ def is_builtin_func(arg):
     otherwise return the arg
     """
     return _builtin_table.get(arg, arg)
+
+
+def fill_missing_names(names):
```
Comment (jreback) [inline pandas/core/common.py:609]: can you type in and out here

**Review by jreback — CHANGES_REQUESTED** [review]
small comments, not 100% sold on the location (maybe @jbrockmendel can think of a better one), but ok. ping on green.

**Comment by jbrockmendel** [comment]
test location makes sense to me

File: doc/source/whatsnew/v1.4.0.rst (hunk 3 above)
Comment (johnzangwill) [reply doc/source/whatsnew/v1.4.0.rst:859]: > can you add backticks around the `n` and `level_n`

Done

File: pandas/core/common.py (hunk 4 above)
Comment (johnzangwill) [reply pandas/core/common.py]: Done, and added version and more docstring.

**Comment by johnzangwill** [comment]
> not 100% sold on the location (maybe @jbrockmendel can think of a better one)

I suspect that @jreback was referring to the fact that I factored the logic into core/common.py.
It seemed a good place to me. My method is ref'ed in:
- core/frame.py
- core/indexes/multi.py
- io/pytables.py
- io/sql.py
- io/json/_table_schema.py

and core/common is already imported as `com` into every one.
But if you have a better suggestion...

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @johnzangwill very nice!

**Pull request merged**